test_util/test_upgrade_vpc.py

```diff
@@ -360,6 +379,104 @@ def test_app_dns_survive_upgrade():
             )
             self.log_test("test_upgrade_vpc.test_app_dns_survive_upgrade", test_app_dns_survive_upgrade)
 
+    def mesos_metrics_snapshot(self, cluster, host):
+        raise NotImplementedError()
+
+    def upgrade_host(self, tunnel, role, bootstrap_url, upgrade_script_path):
+        # Download the upgrade script for the new DC/OS.
+        tunnel.remote_cmd(curl_cmd + ['--remote-name', upgrade_script_path])
+
+        # Upgrade to the new DC/OS.
+        tunnel.remote_cmd(['sudo', 'bash', 'dcos_node_upgrade.sh'])
+
+    @retry(
+        wait_fixed=(1000 * 5),
+        stop_max_delay=(1000 * 60 * 5),
+        # Retry on SSH command error or metric not equal to expected value.
+        retry_on_exception=(lambda exc: isinstance(exc, CalledProcessError)),
+        retry_on_result=(lambda result: not result),
+    )
+    def wait_for_mesos_metric(self, cluster, host, key, value):
+        """Return True when host's Mesos metric key is equal to value."""
+        return self.mesos_metrics_snapshot(cluster, host).get(key) == value
+
+    def upgrade_dcos(self, cluster, installer_url, version, add_config_path=None):
```

Comment:
@branden how do you see this being useful elsewhere? As it stands, the entire cluster.py file will only be used by the upgrade test after dcos-launch is used to drive the CI, and even then, its only purpose for those tests now is having the shared constructor for running tests. What I am thinking for the long term is that cluster will lose all the provisioning and cloudformation bits, and become an abstraction whose main purpose is running SSH commands against roles. E.G. I don't need to say "do this to this host" but rather "do this to all agents in parallel or serial"

The goal being that many people have wanted to write tests to perform remote commands and so this would be one interface. There might be an "upgrader" module after that which combines the api_session objects with the ssh controller.

What do you think?